python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,726 @@
+            raise e
+
+    def set_cloudinit_userdata(self, vm_id):
+        logger.info("Setting cloudinit userdata for vm {}".format(vm_id))
+
+        metadata = '{"cloud_name": "vSphere"}'
+
+        # Read the public key that was generated previously.
+        with open(PUBLIC_KEY_PATH, "r") as file:
+            public_key = file.read().rstrip("\n")
+
+        # This file contains the userdata with default values.
+        # We want to add the key that we generate with create_key_pair
+        # function into authorized_keys section
+
+        f = open(USER_DATA_FILE_PATH, "r")
+        data = yaml.load(f, Loader=yaml.FullLoader)
+        for _, v in data.items():
+            for user in v:
+                if isinstance(user, dict):
+                    user["ssh_authorized_keys"] = [public_key]
+        f.close()
+
+        modified_userdata = yaml.dump(data, default_flow_style=False)
+
+        # The userdata needs to be prefixed with #cloud-config.
+        # Without it, the cloudinit spec would get applied but
+        # it wouldn't add the userdata on the VM.
+
+        modified_userdata = "#cloud-config\n" + modified_userdata
+        logger.info("Successfully modified the userdata file for vm {}".format(vm_id))
+
+        # Create cloud-init spec and apply
+        cloudinit_config = CloudinitConfiguration(
+            metadata=metadata, userdata=modified_userdata
+        )
+        cloud_config = CloudConfiguration(
+            cloudinit=cloudinit_config, type=CloudConfiguration.Type("CLOUDINIT")
+        )
+        config_spec = ConfigurationSpec(cloud_config=cloud_config)
+        global_dns_settings = GlobalDNSSettings()
+        adapter_mapping_list = []
+        customization_spec = CustomizationSpec(
+            configuration_spec=config_spec,
+            global_dns_settings=global_dns_settings,
+            interfaces=adapter_mapping_list,
+        )
+
+        # create customization specification by CustomizationSpecs service
+        specs_svc = self.vsphere_sdk_client.vcenter.guest.CustomizationSpecs
+        spec_name = str(uuid.uuid4())
+        spec_desc = (
+            "This is a customization specification which includes"
+            "raw cloud-init configuration data"
+        )
+        create_spec = specs_svc.CreateSpec(
+            name=spec_name, description=spec_desc, spec=customization_spec
+        )
+        specs_svc.create(spec=create_spec)
+
+        vmcust_svc = self.vsphere_sdk_client.vcenter.vm.guest.Customization
+        set_spec = vmcust_svc.SetSpec(name=spec_name, spec=None)
+        vmcust_svc.set(vm=vm_id, spec=set_spec)
+
+        logger.info("Successfully added cloudinit config for vm {}".format(vm_id))
+
+    # Example: If a tag called node-status:initializing is present on the VM.
+    # If we would like to add a new value called finished with the node-status key.
+    # We'll need to delete the older tag node-status:initializing first before creating
+    # node-status:finished
```

Comment:
Interesting, is there no way to just overwrite a tag's value? Or is there some other benefit to removing the older tag first?

Reply:
In vSphere the tag format is string, not dict.
There is no notion of "tag's value" in vSphere.
So we must either:
1. remove the old tag first then add the new tag
2. add the new tag first then remove the old tag

There should be no update operation in vSphere.